Zend/zend.c

```diff
@@ -1176,9 +1176,19 @@ ZEND_API ZEND_COLD ZEND_NORETURN void _zend_bailout(const char *filename, uint32
 		exit(-1);
 	}
 	gc_protect(1);
+
+	if (CG(in_compilation)) {
+		CG(in_compilation) = 0;
+		/* We bailout during compilation which may for example leave stale entries in CG(loop_var_stack).
+		 * If code is compiled during shutdown, we need to make sure the compiler is reset to a clean state,
+		 * otherwise this will lead to incorrect compilation during shutdown.
+		 * We don't do a full re-initialization via init_compiler() because that will also reset streams and resources. */
+		shutdown_compiler();
```

Comment:
Other calls to `shutdown_compiler` are wrapped in `zend_try`, do we need that here? I don't believe so, looking at its implementation. But maybe you can double-check. :slightly_smiling_face:

Reply:
I don't really know why other locations surround it in a try, it doesn't seem necessary.
However, if wanted I can do this, it doesn't seem harmful.

Reply:
I checked again, and none of the frees in `shutdown_compiler` can throw, because they are either just calls to efree, or free hash tables that contain strings or pointers.

Reply:
I'm not sure if we call ``shutdown_function`` after a fatal error on purpose. This is definitely not safe.
We especially disable destructor calls in ``php_error_cb()``.

In any case, I think it's better to move this recovery code into ``php_error_cb()``.

Reply:
@dstogov Can you clarify why this would be problematic for shutdown functions? When compiling new files through includes in shutdown, the compile structures are cleaned and newly initialized. Nothing should be referencing the old values either, as the compilation has previously bailed. I might be missing something.

Reply:
> > I'm not sure if we call shutdown_function after a fatal error on purpose. This is definitely not safe.
> > We especially disable destructor calls in php_error_cb().
> 
> Good question, I don't know but I wouldn't be surprised people depend on this.

This is right. People depends on every weird feature :(
This error handler may be called after ANY fatal error that stood something in inconsistent state. E.g. after memory overflow.

Reply:
> @dstogov Can you clarify why this would be problematic for shutdown functions? When compiling new files through includes in shutdown, the compile structures are cleaned and newly initialized. Nothing should be referencing the old values either, as the compilation has previously bailed. I might be missing something.

I don't claim your recovery code. Recovery after parse errors should be possible.
I ask you to move this code to a more appropriate place (into ``php_error_cb().`` or even higher).
Also it may make sense to do this only for ``E_PARSE`` errors.

Reply:
The error from the test case is an `E_COMPILE_ERROR` instead of a `E_PARSE`.
I adjusted the code in php_error_cb to recover the compiler if either of these two happen.